rslib/src/import_export/package/apkg/import/mod.rs

```diff
@@ -113,11 +115,15 @@ fn collection_to_tempfile(meta: &Meta, archive: &mut ZipArchive<File>) -> Result
     let mut zip_file = archive.by_name(meta.collection_filename())?;
     let mut tempfile = NamedTempFile::new()?;
     if meta.zstd_compressed() {
-        copy_decode(zip_file, &mut tempfile)
+        copy_decode(&mut zip_file, &mut tempfile)
     } else {
         io::copy(&mut zip_file, &mut tempfile).map(|_| ())
     }
-    .map_err(|err| AnkiError::file_io_error(err, tempfile.path()))?;
+    .with_context(|_| FileIoSnafu {
+        path: tempfile.path(),
+        op: FileOp::copy(zip_file.name()),
+    })
+    .map_err(AnkiError::FileIoError)?;
```

Comment:
This feels a little boilerplatey - it would be nice if we could just use with_context(...)? instead. I think we could do that if we make the top-level error derive Snafu and switch to struct enums? https://github.com/ankitects/anki/pull/2030#issuecomment-1226779795.

Reply:
Yes, that's what I meant in caveat 1. However, maybe that original plan isn't such a good idea after all.
You made an argument against inlining error structs in the case of more specific errors, like CardTemplateError. I've realised that it might also be a problem in universal low-level errors like FileIoError: Let's say we have a module that returns its own error, but also works with input, so it must handle io errors. If AnkiError::FileIoError was an inline struct, we couldn't reuse it in the module error enum.
e0fb628d4d2f616a53d947e271fd5359aded890c gets rid of most of the boilerplate.

Reply:
Just to confirm we're on the same page - in the `FutureAnkiError` I proposed above, I was not proposing that the error context be directly inlined into the enum variant. By only using a single `source` field for each struct variant, it's effectively the same as the existing tuple struct approach, and each of the embedded errors can be used separately.  And the #[snafu(...)] line gives us the from Impl with 3 less lines. WDYT?

Reply:
Sorry, I lost track a bit during the hiatus. Of course, your patch above will get rid of the boilerplate anyway and e0fb628d4d2f616a53d947e271fd5359aded890c is superfluous.
I mean this PR only to be a demo and playground. Once we are decided on the basic points (and I guess we are about now?) I will replay the relevant parts, including your patch, on a clean branch. My bad for not clarifying this sooner!

Reply:
No problem, sounds like we agree on the basic points :-)